Code cleanup

SamMousa opened the pull request:
This is a big cleanup PR that fixes a plethora of smaller issues and rewrites other parts.

| Q             | A
| ------------- | ---
| Breaks BC?    | ✔️ big time

SamMousa commented:
Apologies for the lack of specificity but the library was so small that a quick rewrite that breaks BC was more efficient. Hope you have time to review and provide feedback.

Some things this PR does:

- Implement interfaces for message signing en encrypting, something Symfony lacks.
- Implement wrappers for Symfony's signers & encrypters that implement the new interfaces.
- Implement interface `SymfonyMessageWrapperInterface` which contains the `getSymfonyEmail()` function that our `Mailer` needs. This allows us to typehint against the interface not the class.
- Rewrote `Logger` to use the PSR provided trait, allowing us to implement just the log function. Added configurable severity mappings and configurable category.
- Removed `with-` functions in the `Mailer`, Yii components are traditionally mutable and reconfigurable so this pattern is not needed.
- Allow user to set their own `TransportFactory` this allows them to configure things like a custom `HttpClient` but also available transports and which `Dispatcher`, if any, to use.
- Added test coverage, almost all code is now fully covered, except `Message` which sits at ~80%. The proxy wrappers that just contain 2 lines of code like `DkimSignerWrapper` have been excluded from code coverage.

samdark commented on src/Mailer.php:
```diff
@@ -72,47 +73,42 @@ public function getSymfonyMailer(): SymfonyMailer
 
     /**
-     * @return TransportInterface
+     * @deprecated This will become private
```
What do you mean by "will become"?

SamMousa replied:
Hmm that's a remnant of when I thought I'd do a small non-bc-breaking PR ;-) will fix it.

samdark commented on src/SMimeEncrypterWrapper.php:
```diff
@@ -0,0 +1,26 @@
+/**
+ * @codeCoverageIgnore This class is a trivial proxy that requires no testing
+ */
+class SMimeEncrypterWrapper implements SymfonyMessageEncrypterInterface
```
```suggestion
class SMimeMessageEncrypter implements SymfonyMessageEncrypterInterface
```

samdark commented on src/SMimeSignerWrapper.php:
```diff
@@ -0,0 +1,26 @@
+/**
+ * @codeCoverageIgnore This class is a trivial proxy that requires no testing
+ */
+class SMimeSignerWrapper implements SymfonyMessageSignerInterface
```
```suggestion
class SMimeMessageSigner implements SymfonyMessageSignerInterface
```

samdark commented on src/SymfonyMessageEncrypterInterface.php:
```diff
@@ -0,0 +1,12 @@
+
+use Symfony\Component\Mime\Message as Message;
+
+interface SymfonyMessageEncrypterInterface
```
```suggestion
interface MessageEncrypterInterface
```

samdark commented on src/SymfonyMessageSignerInterface.php:
```diff
@@ -0,0 +1,12 @@
+
+use Symfony\Component\Mime\Message;
+
+interface SymfonyMessageSignerInterface
```
```suggestion
interface MessageSignerInterface
```

Krakozaber commented:
Is the completion of the work planned ?

SamMousa commented:
Yes definitely. This branch is already in production so I'm pretty sure it works.
Needs a good review though.

samdark commented:
@SamMousa shouldn't readme be adjusted?

SamMousa commented:
Hmm, good point, very probably yes. Will try to make some time to finish this!

Krakozaber commented:
Is everything bad?

SamMousa commented:
Well it depends on what year we live in. When Yii2 came out it was good, but that was a long time ago. Nowadays the standard "Yii" way is not considered a clean / proper way of writing code. (Not blaming anyone, back then we had php 5.6, no typing, no static analysis etc)
Modern code is different and practices have improved as well.

Krakozaber commented:
What needs to be completed for the merger?

SamMousa commented:
The Readme

Krakozaber commented:
@SamMousa Is there any way I can help with this?

SamMousa commented:
Definitely, just check the current Readme and make sure it's code samples are still correct!

SamMousa commented:
@samdark as far as I can tell the readme is good, since the public configuration API is mostly unchanged.

This means this can be merged. In my opinion: release it as a new major version, it may require changes in consuming code if they use advanced stuff like message signing or encryption.

samdark commented on src/Logger.php:
```diff
@@ -134,30 +53,17 @@ public function debug($message, array $context = []): void
-                break;
-            default:
-                $level = \yii\log\Logger::LEVEL_INFO;
+        if (! is_string($level)) {
```
```suggestion
        if (!is_string($level)) {
```

samdark commented on src/Logger.php:
```diff
@@ -134,30 +53,17 @@ public function debug($message, array $context = []): void
+            throw new \InvalidArgumentException("This logger only supports string levels");
         }
-        Yii::getLogger()->log($message, $level, __METHOD__);
+        if (! isset($this->map[$level])) {
```
```suggestion
        if (!isset($this->map[$level])) {
```

samdark commented on src/Mailer.php:
```diff
@@ -58,158 +62,100 @@ private function createSymfonyMailer(): SymfonyMailer
-            $this->_transport = $transport;
-        } elseif (is_array($transport)) {
-            $this->_transport = $this->createTransport($transport);
+        if (! is_array($transport) && ! $transport instanceof TransportInterface) {
```
```suggestion
        if (!is_array($transport) && !$transport instanceof TransportInterface) {
```

samdark commented on src/Mailer.php:
```diff
@@ -58,158 +62,100 @@ private function createSymfonyMailer(): SymfonyMailer
-        if (!is_object($this->_transport)) {
-            $this->_transport = $this->createTransport($this->_transport);
+        /** @psalm-suppress RedundantPropertyInitializationCheck Yii2 configuration flow does not guarantee full initialisation */
+        if (! isset($this->_transport)) {
```
```suggestion
        if (!isset($this->_transport)) {
```

samdark commented on src/Logger.php:
```diff
@@ -134,30 +53,17 @@ public function debug($message, array $context = []): void
-            default:
-                $level = \yii\log\Logger::LEVEL_INFO;
+        if (! is_string($level)) {
+            throw new \InvalidArgumentException("This logger only supports string levels");
```
```suggestion
            throw new \InvalidArgumentException("This logger only supports string levels.");
```

samdark commented on src/Logger.php:
```diff
@@ -134,30 +53,17 @@ public function debug($message, array $context = []): void
         }
-        Yii::getLogger()->log($message, $level, __METHOD__);
+        if (! isset($this->map[$level])) {
+            throw new InvalidArgumentException("Unknown logging level $level");
```
```suggestion
            throw new InvalidArgumentException("Unknown logging level $level.");
```

samdark commented on src/Mailer.php:
```diff
@@ -58,158 +62,100 @@ private function createSymfonyMailer(): SymfonyMailer
     {
-        if (!($message instanceof Message)) {
-            throw new RuntimeException(sprintf(
+        if (! ($message instanceof MessageWrapperInterface)) {
```
```suggestion
        if (!($message instanceof MessageWrapperInterface)) {
```

samdark commented on src/Message.php:
```diff
@@ -153,122 +192,79 @@ public function setSender(string $address): self
-        } else {
-            $file['name'] = null;
+        if (empty($options['fileName'])) {
+            throw new InvalidConfigException('A valid file name must be passed when embedding content');
```
```suggestion
            throw new InvalidConfigException('A valid file name must be passed when embedding content.');
```

samdark commented on src/Message.php:
```diff
@@ -352,14 +353,10 @@ private function convertAddressesToStrings(array $addresses)
-        foreach ($strings as $address => $name) {
-            if (!is_string($address)) {
+        foreach ((array) $strings as $address => $name) {
+            if (! is_string($address)) {
```
```suggestion
            if (!is_string($address)) {
```

samdark commented on tests/MessageTest.php:
```diff
@@ -14,24 +16,24 @@
         ]);
         $filePath = $this->getTestFilePath();
-        if (!file_exists($filePath)) {
+        if (! file_exists($filePath)) {
```
```suggestion
        if (!file_exists($filePath)) {
```

samdark commented on tests/MessageTest.php:
```diff
@@ -78,9 +80,9 @@ protected function createTestMessage()
+    private function createImageFile($fileName = 'test.jpg', $text = 'Test Image'): string
     {
-        if (!function_exists('imagecreatetruecolor')) {
+        if (! function_exists('imagecreatetruecolor')) {
```
```suggestion
        if (!function_exists('imagecreatetruecolor')) {
```

samdark commented on tests/TestCase.php:
```diff
@@ -34,7 +34,7 @@ protected function mockApplication($config = [], $appClass = '\yii\console\Appli
     {
         $vendor = dirname(dirname(__DIR__)) . '/vendor';
-        if (!is_dir($vendor)) {
+        if (! is_dir($vendor)) {
```
```suggestion
        if (!is_dir($vendor)) {
```

samdark commented on tests/compatibility.php:
```diff
@@ -4,14 +4,17 @@
 namespace PHPUnit\Framework\Constraint {
-    if (!class_exists('PHPUnit\Framework\Constraint\Constraint') && class_exists('PHPUnit_Framework_Constraint')) {
-        abstract class Constraint extends \PHPUnit_Framework_Constraint {}
+    if (! class_exists('PHPUnit\Framework\Constraint\Constraint') && class_exists('PHPUnit_Framework_Constraint')) {
```
```suggestion
    if (!class_exists('PHPUnit\Framework\Constraint\Constraint') && class_exists('PHPUnit_Framework_Constraint')) {
```

samdark commented on tests/compatibility.php:
```diff
@@ -4,14 +4,17 @@
 namespace PHPUnit\Framework {
-    if (!class_exists('PHPUnit\Framework\TestCase') && class_exists('PHPUnit_Framework_TestCase')) {
-        abstract class TestCase extends \PHPUnit_Framework_TestCase {
+    if (! class_exists('PHPUnit\Framework\TestCase') && class_exists('PHPUnit_Framework_TestCase')) {
```
```suggestion
    if (!class_exists('PHPUnit\Framework\TestCase') && class_exists('PHPUnit_Framework_TestCase')) {
```

samdark reviewed (approved):
👍 Need a line for CHANGELOG.

SamMousa commented:
@samdark your code style changes should be enforced by easy coding standard. (Check `ecs.php`) 

I can use a different standard / config, but the idea is to automate the check so that code style feedback by humans isn't needed and any PR will fail if it doesn't abide by the rules.

samdark commented:
I know. Doing that in Yii3 already via StyleCI.

nd4c commented:
I've just quickly reviewed some of the changes.  I'm concerned that getSymfonyMailer() method has been made private.  This makes it impossible to extend the Mailer class with a new sendMessage() method.  Please keep it at least protected.

In fact it would be far nicer if $encryptor, $signer and $dkimSignerOptions were also protected and not private.

nd4c commented:
Please also not that in the sendMessage method, it needs to re-throw the Exception since the calling code will need that to extract the errors.  No need to return false;

nd4c commented:
There are a lot of changes in this.  Is it possible to include a public socket timeout property?

https://github.com/yiisoft/yii2-symfonymailer/issues/40

SamMousa commented:
> In fact it would be far nicer if $encryptor, $signer and $dkimSignerOptions were also protected and not private.

They are public as far as I can tell. So you can configure them however you like.

SamMousa commented:
> There are a lot of changes in this. Is it possible to include a public socket timeout property?

This is not really related to these changes. But the way to go about it (and this is purely the Symfony part of the library) is as you suspected.
You'd use a wrapper class (composition instead of inheritance) since Symfony even goes so far as to make their classes final so you cannot extend them.

```php
class SmtpTransportWithTimeoutFactory implements TransportFactoryInterface {
    public function __construct(private TransportFactoryInterface $factory, private float $timeout)
    {
    }

    public function create(Dsn $dsn): TransportInterface
    {
        $result = $this->factory->create($dsn);
        if ($result instanceof SmtpTransport) {
            $result->getStream()->setTimeout($this->timeout);
        }
        return $result;
    }

    public function supports(Dsn $dsn): bool {
        return $this->factory->supports($dsn);
    }
}

```

SamMousa commented:
@samdark I think it's really done now!

nd4c commented:
I really hate to bother you guys. I am really looking forward to these changes

When I call:
`$mailer = Yii::$app->mailer;`
I keep getting a error:

Error 
Typed property ....\Mailer::$transportFactory must not be accessed before initialization

I was trying to set it up so I could start building and wiring up a Transport.  The config is very normal and complies with the ReadMe sample with the exception that the 'dsn' key is not used as this will override the other settings.

    private function createTransport(array $config = []): TransportInterface
    {
        if (array_key_exists('dsn', $config)) {
            $transport = $this->transportFactory->fromString($config['dsn']);
        } elseif (array_key_exists('scheme', $config) && array_key_exists('host', $config)) {
            $dsn = new Dsn(
                $config['scheme'],
                $config['host'],
                $config['username'] ?? '',
                $config['password'] ?? '',
                $config['port'] ?? null,
                $config['options'] ?? [],
            );
            $transport = $this->transportFactory->fromDsnObject($dsn);
        } else {
            throw new InvalidConfigException('Transport configuration array must contain either "dsn", or "scheme" and "host" keys.');
        }
        return $transport;
    }

I have tried different ways of "installing" the new version - automagical loading can get confusing - but with the config containing a transport configuration array it appears that the BaseObject constructor is calling Yii::configure($this, $config) before calling $this->init();

Yii::configure will end up calling setTransport with the configuration array.  And since it's an array setTransport calls createTransport which tries to access $this->transportFactory and that causes an exception:

Typed property ....\Mailer::$transportFactory must not be accessed before initialization

It very possible that I've wired up the Mailer wrong - if that's the case I'm very sorry for wasting your time.

nd4c commented:
This is a sequential view of what's happening, starting with BaseObject::_construct().

yii\base\BaseObject calls BaseYii::configure() before calling $this->init()

    public function __construct($config = [])
    {
        if (!empty($config)) {
            Yii::configure($this, $config);  [THE CALL]
        }
        $this->init();
    }
 
BaseYii calls yii\base\Component::_set()

    public static function configure($object, $properties)
    {
        foreach ($properties as $name => $value) {
            $object->$name = $value; 
 
Component calls Mailer::setTransport because we have a 'transport' => array configuration.

    public function __set($name, $value)
    {
        $setter = 'set' . $name;
        if (method_exists($this, $setter)) {
            // set property
            $this->$setter($value);
 
Mailer calls its own createTransport because $transport is a configuration array

    public function setTransport($transport): void
    {
        $this->_transport = $transport instanceof TransportInterface ? $transport : $this->createTransport($transport);
 
And createTransport tries to access $this->transportFactory before it's initialized by init()

        } elseif (array_key_exists('scheme', $config) && array_key_exists('host', $config)) {
            $dsn = new Dsn(
                $config['scheme'],
                $config['host'],
                $config['username'] ?? '',
                $config['password'] ?? '',
                $config['port'] ?? null,
                $config['options'] ?? [],
            );
            $transport = $this->transportFactory->fromDsnObject($dsn);

SamMousa commented:
You must be in a different timezone 😅 I'll have a look. Note that the tests should show an example of how I wired it up. That is not to say this isn't a bug.

SamMousa commented:
@nd4c give it another go; I've fixed it and added a test case for it.
Funny thing is that this type of bug is normally caught by static analysis, but we had to suppress those initialization checks because of the way object creation works in the framework :-/

SamMousa commented:
Hmm, should we even support these custom options?

Instead we could support just DSN and reference this doc: https://symfony.com/doc/current/mailer.html#transport-setup

nd4c commented:
I would recommend allowing them because if someone creates a new Transport that has a lot of 'options' it's much easier to see what they are rather than appending them to the 'dsn' string.

nd4c commented:
Also, I would recommend using 'protected' instead of 'private'.  This allows the classes to be extended.

Quote from the "About Yii" section of the "Guide":
`Yii is extremely extensible. You can customize or replace nearly every piece of the core's code. You can also take advantage of Yii's solid extension architecture to use or develop redistributable extensions.`

And since this is an official Yii extension rather than an independent library it makes sense to me to follow that "extensible" aspect.

SamMousa commented:
>  This allows the classes to be extended.

You keep coming back to this, but I believe it is a fundamental misunderstanding of how to write extensible code. Code being extensible does not have to mean that classes are extendable.
Anything you have mentioned wanting to do by using custom configuration or extending classes you can do better by using composition.

> I would recommend allowing them because if someone creates a new Transport that has a lot of 'options' it's much easier to see what they are rather than appending them to the 'dsn' string.

The library we use uses a DSN though, and the DSN object inherently loses features like `failover` or `roundrobin` configuration. Which means that we sacrifice features from the parent library just to introduce our own configuration syntax which we then need to document and help users with. 

Thinking it through a bit more I'm pretty sure that even leaving in the `transport` configuration option is not the best design. 
Ideally you configure the factory and have it support your custom transport. The custom transport is than configured using a DSN like all other transports already supported by the library.

This is way cleaner:
```php
[
    'class' => Mailer::class,
    'dsn' => "sendgrid+api://{$env->get('SENDGRID')}@default"
]
```
And if you need a custom factory implement that:

```php
[
    'class' => Mailer::class,
    'transportFactory' => new Transport(new MyTransportFactory()),
    'dsn' => "my-custom-protocol://{$env->get('SENDGRID')}@default"
]
```

nd4c commented:
Composition it is then.  Of course, you realize that the Mailer class extends BaseMailer...

I don't see that strictly using a 'dsn' element is necessarily cleaner.  When using certain Transports, yes, it's cleaner.  But not for all Transports.

	'transport' => [
		'scheme' => 'smtps',
		'host' => 'mail.longwinded.com',
		'port' => 2525,
		'username' => 'customersupport@mailingstation.anotherverylongurl.com',
		'password' => 'We prefertousea Verylongandunforgettablepassphrase insteadof a password',
		'options' => [
			'verify_peer' => 0,
			'local_domain' => 'example.org',
			'restart_threshold' =< 10,
			'restart_threshold_sleep' => 1,
			'ping_threshold' => 200,
		],
	],

That would be a very long 'dsn' string.  And, again, the DSN object just parses the string to get back to what's clearly displayed in the array.

nd4c commented:
I guess if I wanted to use `failover` or `roundrobin` with such complicated transport settings I would probably need a Mailer Factory that could automatically compose the 'dsn' string from a much more readable config array.  So it might make sense to just use the 'dsn' string.  I just keep feeling that we'd be missing a nice feature.

SamMousa commented:
> Composition it is then. Of course, you realize that the Mailer class extends BaseMailer...

I do, there's no easy way around it. Also realize that the framework is +- 10 years old, it's core code is not a best practice in 2022.

Krakozaber commented:
[samdark](https://github.com/samdark) added this to the [3.0.0](https://github.com/yiisoft/yii2-symfonymailer/milestone/1) milestone [9 days ago](https://github.com/yiisoft/yii2-symfonymailer/pull/27#event-7318282013)

does that mean never?

samdark commented:
No. That means it will be a major release of this exact extension.

machour commented:
@SamMousa thank you for reworking this!

I'm wondering if we could also include [`MailerEvents`](https://symfony.com/doc/current/mailer.html#mailer-events) as Yii events? (definitely needed to inline CSS or stuff like this prior to sending the message)

SamMousa commented:
> @SamMousa thank you for reworking this!
> 
> I'm wondering if we could also include [`MailerEvents`](https://symfony.com/doc/current/mailer.html#mailer-events) as Yii events? (definitely needed to inline CSS or stuff like this prior to sending the message)

Not in this pr. Let's do that after maybe

tgerakitis commented:
just curious, what is the ETA on this PR? 
or is it possible to require this branch using composer?

thanks for your hard work! 😊

SamMousa commented:
> or is it possible to require this branch using composer?

Yes. Google is your friend.

samdark commented:
@SamMousa sorry it took so long. Would you please help closing issues this PR solves? Thanks.

Krakozaber commented:
@samdark need a tag

samdark commented:
@SamMousa, @Krakozaber is readme alright?

nd4c commented:
Also, I believe the 'scheme' settings is better to be left out.  

'scheme' => 'smtps' is kind of misleading.  It doesn't exactly force ssl, but rather causes the code to choose a different transport.

I forget the details, but it's the 'port' => '465' that cause the code to use ssl.  And the default transport works better than the 'smtps' transport.

So the better config would be:

```
'transport' => [
                'host' => '',
                'username' => '',
                'password' => '',
                'port' => 465,
            ],

```

nd4c commented:
I've never dones a PR before.  Any tips?

SamMousa commented:
Personally I believe you should be using DSN in all cases. The underlying library uses DSN and by not using it all you do is lose functionality.

> I've never dones a PR before. Any tips?

For these simple things you can just do them inside Github and it will do most of the meta-work for you. 
Go here: https://github.com/yiisoft/yii2-symfonymailer/blob/master/README.md and click the pencil top right.

Good luck!

nd4c commented:
Actually the underlying library uses the inidividual fields.  It will convert any DSN string to individual attributes on it's DSN object.  And not all transports are easy to for a developer to "read" as a DSN string.  (And for that matter, the DSN string is not technically a DSN, but rather a list of configuration settings...)

I will work on the README.md file.
